pkg/tfshim/sdk-v2/provider2.go

```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
 	}
 	return dd
 }
+
+func normalizeBlockCollections(val cty.Value, res *schema.Resource) cty.Value {
+	// Full rules about block vs attr
+	//nolint:lll
+	// https://github.com/hashicorp/terraform-plugin-sdk/blob/1f499688ebd9420768f501d4ed622a51b2135ced/helper/schema/core_schema.go#L60
+	sch := res.CoreConfigSchema()
+	if !val.Type().IsObjectType() {
+		contract.Failf("normalizeBlockCollections: Expected object type, got %v", val.Type().GoString())
+	}
+	valMap := val.AsValueMap()
+
+	for fieldName := range sch.BlockTypes {
+		if !val.Type().HasAttribute(fieldName) {
+			continue
+		}
+		subVal := val.GetAttr(fieldName)
+		if subVal.IsNull() {
+			fieldType := val.Type().AttributeType(fieldName)
+			// Only lists and sets can be blocks and pass InternalValidate
+			// Ignore other types.
+			if fieldType.IsListType() {
+				glog.V(10).Info("normalizing list %s, %s", fieldName, fieldType.ElementType())
```

Comment:
Nit: normalizing is a bit of a generic term, may say something like "replacing a nil list with an empty list because the underlying TF property is a block"